Mathlib/Tactic/Algebraize.lean

```diff
@@ -0,0 +1,298 @@
+which it tries to add the corresponding `Algebra` properties to the context. It only looks for
+properties that have been tagged with the `algebraize` attribute, and uses this tag to find the
+corresponding `Algebra` property. -/
+def addProperties (t : Array Expr) : TacticM Unit := withMainContext do
+  let ctx ← getLCtx
+  ctx.forM fun decl => do
+    if decl.isImplementationDetail then return
+    let (nm, args) := decl.type.getAppFnArgs
+    -- Check if the type of the current hypothesis has been tagged with the `algebraize` attribute
+    match Attr.algebraizeAttr.getParam? (← getEnv) nm with
+    -- If it has, `p` will be the name of the corresponding `Algebra` property (or a constructor)
+    | some p =>
+      -- The last argument of the `RingHom` property is assumed to be `f`
+      let f := args[args.size - 1]!
+      -- Check that `f` appears in the list of functions given to `algebraize`
+      if ¬ (← t.anyM (Meta.isDefEq · f)) then return
+
+      let cinfo ← getConstInfo p
+      let n ← getExpectedNumArgs cinfo.type
+      let pargs := Array.mkArray n (none : Option Expr)
+      /- If the attribute points to the corresponding `Algebra` property itself, we assume that it
+      is definitionally the same as the `RingHom` property. Then, we just need to construct its type
+      and the local declaration will already give a valid term. -/
+      match cinfo with
+      | .inductInfo _ =>
+        let pargs := pargs.set! 0 args[0]!
+        let pargs := pargs.set! 1 args[1]!
+        let tp ← mkAppOptM p pargs -- This should be the type `Algebra.Property A B`
+        unless (← synthInstance? tp).isSome do
+        liftMetaTactic fun mvarid => do
+          let nm ← mkFreshBinderNameForTactic `algebraizeInst
+          let mvar ← mvarid.define nm tp decl.toExpr
+          let (_, mvar) ← mvar.intro1P
+          return [mvar]
+      /- Otherwise, the attribute points to a constructor of the `Algebra` property. In this case,
+      we assume that the `RingHom` property is the last argument of the constructor (and that
+      this is all we need to supply explicitly). -/
+      | .ctorInfo ctor =>
+        -- construct the desired value
+        let pargs := pargs.set! (n - 1) decl.toExpr
+        let val ← mkAppOptM p pargs
+
+        -- construct the expected type
+        let alg ← mkAppOptM ``Algebra #[args[0]!, args[1]!, none, none]
+        let algInst := (← synthInstance? alg)
+        let mut argsType := Array.mkArray (ctor.numParams) (none : Option Expr)
+        argsType := argsType.set! 0 args[0]!
+        argsType := argsType.set! 1 args[1]!
+        argsType := argsType.set! (ctor.numParams - 1) algInst
+        let tp := ← mkAppOptM ctor.induct argsType
+
+        unless (← synthInstance? tp).isSome do
+        liftMetaTactic fun mvarid => do
+          let nm ← mkFreshBinderNameForTactic `algebraizeInst
+          let mvar ← mvarid.define nm tp val
```

Comment:
I realised now that this change I made wouldn't be necessary: since all ring properties are always `Prop` (at least so far), I think it doesn't matter if one uses `mvarid.define` or `mvarid.note` in `addProperties` due to proof-irrelevance.

So this construction of the expected type above here could be removed again

Reply:
I don't think this is a critical issue. So let's fix it in a follow-up PR. Because this tactic is going to be very helpful in the immediate future.